### ISSUE TITLE
T-SQL: Semi-colon unparseable for various statements

### ISSUE THREAD

**Issue by skykasko** [sqlfluff/sqlfluff#2029]
## Expected Behaviour

SQLFluff should parse [USE][] statements in T-SQL that end with a semicolon.

[USE]: https://docs.microsoft.com/en-us/sql/t-sql/language-elements/use-transact-sql

## Observed Behaviour

```shell
$ sqlfluff lint --dialect tsql use.sql
== [use.sql] FAIL
L:   1 | P:  23 |  PRS | Line 1, Position 23: Found unparsable section: ';'
All Finished 📜 🎉!
```

## Steps to Reproduce

Create `use.sql`, taken from the [example][] in the Microsoft documentation:

```sql
USE AdventureWorks2012;
GO

```

Run `sqlfluff lint --dialect tsql use.sql`.

Note that no parsing error is reported if the semicolon is removed. The presence or absence of the `GO` command makes no difference.

[example]: https://docs.microsoft.com/en-us/sql/t-sql/language-elements/use-transact-sql#examples

## Dialect

`tsql`

## Version

- SQLFluff: 0.8.2
- Python: 3.9.9

**Comment by jpy-git** [comment]
Think this is a bigger issue as something like `DROP SCHEMA sdfa;` also won't parse at the semi-colons. I imagine there are a lot more, especially for grammar inherited from ANSI.

The semi-colon implementation is a bit weird for T-SQL.

For every other dialect the statement terminator `;` is on the same parse depth as the statement segment itself:
![image](https://user-images.githubusercontent.com/80432516/144527446-426261f0-4f09-4296-bac9-3f6594fdd0d9.png)
and seems to be generic and not referenced in any of the statements themselves.

However, for T-SQL it seems to be at a random depth in the parse tree (same level as FROM in selectable):
![image](https://user-images.githubusercontent.com/80432516/144527572-eda6bfa8-4afc-44d3-8f07-a2ddb1eae1bb.png)
and is explicitly referenced in the grammar in various places:
![image](https://user-images.githubusercontent.com/80432516/144527698-3216b8cc-06cb-4f70-91ef-d0789f934d65.png)

Seems to me like the semi-colon segment in T-SQL should be similar to ANSI and parsed at the same level as StatementSegment.

**Comment by jpy-git** [comment]
Initial thoughts are the BatchSegment T-SQL needs to be treated more similarly to FileSegment in ANSI.

![image](https://user-images.githubusercontent.com/80432516/144528139-d63293ea-8c9f-48c4-b8f8-b49f99689de0.png)
And then remove the random references to DelimiterSegment in the various T-SQL grammars in favour of this one.

**Comment by jpy-git** [comment]
I also discussed this recently in #1979 (https://github.com/sqlfluff/sqlfluff/pull/1979#issuecomment-979546395) as making the T-SQL semi-colon logic more consistent would allow me to better refine L052 as well 👍

**Comment by skykasko** [comment]
Thanks for the explanation. It sounds like it may be related to this SQLite bug, which also involves a semicolon at an unexpected depth: <https://github.com/sqlfluff/sqlfluff/issues/1846#issuecomment-963077607>.

**Comment by jpers36** [comment]
Since semicolons are optional in TSQL, using Delimited() at the BatchSegment/FileSegment level is a no-go.  At least, I couldn't figure out a way for it to work, so I had to rewrite the dialect from the top level to get any sort of functionality out of it.  This did force me to add optional Delimiter refs at the end of statements. I believe this to be the only solution, but I admit the actual implementation was both piecemeal and inconsistent.

**Comment by jpy-git** [comment]
@jpers36 makes sense, maybe it'd be worth exploring a new grammar segment: OptionallyDelimited

Basically any number of the segment optionally with a delimiter between them (sort of a mix between AnyNumberOf and Delimited)?
```python
OptionallyDelimited(Ref("StatementSegment"), delimiter=Ref("DelimiterSegment"))
```
allowing
```
SELECT foo FROM bar

SELECT foo FROM bar

SELECT foo FROM bar;

SELECT foo FROM bar
```

Worth considering if it's possible 😄

**Comment by skykasko** [comment]
Closing as I can no longer reproduce any of the parsing errors mentioned in this issue when running SQLFluff 0.9.1. My best guess would be this was resolved in https://github.com/sqlfluff/sqlfluff/pull/2124, though I'm not sure.
